torch/fx/passes/runtime_assert.py

```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
-            ras_exprs.add(ras.expr)
-    ras_by_symbol[None] = new_ras  # type: ignore[index]
-
     # We are going to mutate the dict
     symbol_to_proxy: Dict[sympy.Symbol, fx.Proxy] = {}
     placeholders = set()
-    last_placeholder = None
+    first_non_placeholder = next(iter(graph.nodes))
     for node in graph.nodes:
         if node.op != "placeholder":
+            first_non_placeholder = node
             break
-        last_placeholder = node
-        placeholders.add(node)
-    if last_placeholder is None:  # no placeholders, just insert before first node
-        last_placeholder = next(iter(graph.nodes))
+        else:
+            placeholders.add(node)
+
+    def _is_intermediate_tensor_shape_call(node: fx.Node) -> bool:
+        """
+        If a size/stride/storage offset call on an intermediate tensor,
+        we can try to compute the value from input shapes instead.
+        """
+        val = None
+        if (
+            node.target in (
+                torch.ops.aten.sym_size.int,
+                torch.ops.aten.sym_stride.int,
+                torch.ops.aten.storage_offset.default,
+            )
+            and node.args[0].op != "placeholder"  # type: ignore[union-attr]
+        ):  # export: sym_size.int(node, dim)
+            val = _get_sym_val(node)
+        if (
+            node.target == operator.getitem
+            and node.args[0].target in (
+                "size",
+                "stride",
+                "storage_offset",
+            )
+            and node.args[0].args[0].op != "placeholder"
+        ):  # dynamo export: getitem(size(node), dim)
+            val = _get_sym_val(node)
+        return (
+            val is not None
+            and not isinstance(val, sympy.Number)
+        )
 
     # Identify what symbols we need to reify.  This isn't strictly needed
     # but helps reduce churn on the graph
     needed_symbols: Set[sympy.Symbol] = set()
     for ras in ras_by_symbol.values():
         for ra in ras:
             needed_symbols.update(free_symbols(ra.expr))
+    for node in gm.graph.nodes:
+        if (
+            node.op == "call_function"
+            and _is_intermediate_tensor_shape_call(node)
+        ):  # use symbols for intermediate shape calls, even if not in runtime asserts
+            sym_expr = _get_sym_val(node)
+            needed_symbols.update(sym_expr.free_symbols)
 
     log.debug("needed_symbols = %s", needed_symbols)
 
+    # Track asserts/checks we've added
+    added_asserts: Set["sympy.Expr"] = set()
+    constrained_unbacked_symbols: Set["sympy.Symbol"] = set()
+
+    def _sympy_interp(symbol_to_proxy, expr):
+        # sympy_interp() with hash consing
+        from sympy import Integer, Number, Symbol
+        from sympy.logic.boolalg import BooleanAtom
+        from torch.utils._sympy.interp import _run_sympy_handler, sympy_interp
+
+        # base cases, don't cache
+        if isinstance(expr, (Integer, Number, Symbol, BooleanAtom)):
```

Comment:
To be clear, the symbol "caches" because the handler just consults symbol_to_proxy. Not sure what happens with the numbers but I guess they get inlined as literals into FX so they don't need to be hash consed.